### ISSUE TITLE
validate_uniqueness_of fails when scoped to several keys

### ISSUE THREAD

**Issue by kiela** [thoughtbot/shoulda-matchers#670]
Recently I came across and issue while I wanted to write tests for validating uniqueness of column scoped to several other columns.

In our project we have `Category` model which has many rules:
```ruby
class Category < ActiveRecord::Base
	has_many :rules, inverse_of: :category
end
```
Our `Rule` model has keyword attribute, belongs to `Category` and also belongs to `RuleType`:
```ruby
class Rule < ActiveRecord::Base
	belongs_to :category, inverse_of: :rules
	belongs_to :rule_type, inverse_of: :rules

	validates :keyword, presence: true, uniqueness: { scoped_to: [:category_id, :rule_type_id] }
end
```
and finally `RuleType` model has type and many rules:
```ruby
class RuleType < ActiveRecord::Base
	has_many :rules, inverse_of: :rule_type
end
```
Tests for `Category` model and `RuleType` model are green but one test for `Rule` model keeps failing on checking the scoped uniqueness. I'm trying to test that thing in this way:
```ruby
require 'rails_helper'

RSpec.describe Rule, type: :model do
	describe 'validations', :focus do
		it { is_expected.to validate_presence_of(:keyword) }
		it {
			FactoryGirl.create :rule
			is_expected.to validate_uniqueness_of(:keyword).scoped_to([:category_id, :rule_type_id])
		}
	end
end
```
and I got this error:
```
Failures:

  1) Rule validations should require case sensitive unique value for keyword scoped to category_id, rule_type_id
     Failure/Error: is_expected.to validate_uniqueness_of(:keyword).scoped_to([:category_id, :rule_type_id])
       Did not expect errors to include "has already been taken" when keyword is set to "RZSanKRMma",
       got errors:
       * "has already been taken" (attribute: keyword, value: "RZSanKRMma") (with different value of category_id)
     # ./spec/models/rule_spec.rb:8:in `block (3 levels) in <top (required)>'
```

I tried to dig a little bit into it and found out that it could be because scopes are checked separately. It means that the test checked if uniqueness of `:keyword` was scoped to `:category_id` and failed when it was supposted to check uniqueness scoped to `:category_id` and `:rule_type_id` together.

**Comment by mcmire** [comment]
This is odd. It looks like your validation and use of the matcher is correct, so I'm at a loss as to why this would be failing. The error message is a little weird but basically this is saying that if you have two records like this:

``` ruby
Rule.create!(keyword: 'some value', category_id: 1, rule_type_id: 1)
Rule.create!(keyword: 'some value', category_id: 2, rule_type_id: 1)
```

you would expect the second record to be created, but fails validation for some reason.

Does your Rule model have some kind of callback that prevents `category_id` from being set to something? Is there anything else out of the ordinary going on in this model?

**Comment by kiela** [comment]
@mcmire Correct.

I have one record and want to create another with with same value for `:keyword` but different value for `:category_id` or `:rule_type_id`. 
I don't have anything in models beside validations and declaration for associations. I have created demo application with tests for that scenario: https://github.com/kiela/validate_scoped_uniqueness
Maybe that would be helpful for investigating that issue.

**Comment by kiela** [comment]
This morning I tested that one more time but using just models and play with them in Rails console. It also failed. I started digging harder what was going on there and it turned out that parameters for scope were ignored because I declared the validation wrong. The parameter for model validation should be `scope: [..]` instead of `scoped_to: [..]` :bowtie:

Anyways, it's fixed now and the issue can be closed :wink:
